addons/hr_expense/models/product_product.py

```diff
@@ -24,3 +24,13 @@ def _compute_standard_price_update_warning(self):
                             "There are unposted expenses linked to this category. Updating the category cost will change expense amounts. "
                             "Make sure it is what you want to do."
                         )
+
+    def write(self, vals):
+        if 'standard_price' in vals:
+            expenses = self.env['hr.expense'].sudo().search([('product_id', 'in', self.ids), ('state', 'in', ['reported', 'draft'])])
+            for expense in expenses:
+                expense.sudo().write({
+                    'product_has_cost': expense.product_id and not expense.company_currency_id.is_zero(expense.product_id.standard_price),
+                    'product_has_tax': bool(expense.product_id.supplier_taxes_id.filtered_domain(self.env['account.tax']._check_company_domain(expense.company_id))),
+                })
+        return super().write(vals)
```

Comment:
expenses env is sudoed so when looping inside, expense is sudoed too
I recommend suffixing a sudoed recordset with _sudo, so we don't risk setting it loose outside of the function